ir/instr.cpp

```diff
@@ -295,6 +295,71 @@ expr UnaryOp::getTypeConstraints(const Function &f) const {
          move(instrconstr);
 }
 
+void TernaryOp::print(ostream &os) const {
+  const char *str = nullptr;
+  switch (op) {
+  case FShl:
+    str = "fshl ";
+    break;
+  case FShr:
+    str = "fshr ";
+    break;
+  }
+
+  os << getName() << " = " << str << A << ", " << B << ", " << C;
+}
+
+expr TernaryOp::funnelShiftNotPoison(const expr &a, const expr &ap,
```

Comment:
are you sure about the semantics here?
Anything that is not simply "ap && bp && cp" looks very suspicious to me. The reason is that 1) it makes it harder to transform funnel shifts to other instructions (e.g. fshl -> shl) since other instructions don't have conditional poison (only select).

Reply:
I'm not sure, and i thought i have listed that as a question in the PR's description,
but apparently not.
On the first look this *seems* more correct - if we **know** we won't use any of
the bits of the value `a` (or `b`), (and since we know that, we **already** _should_
be able to transform into normal shift), should the fact that that unused value
is `poison` matter?
But i can dumb this down into the usual `ap && bp && cp`,
or start a llvm-dev thread with this question too.

Reply:
From my experience, having short-circuiting definitions of poison doesn't usually work. The reason is what I listed about: converting between instructions then becomes tricky because some instructions will have short-circuit for some cases and others won't, causing a mismatch (and hence forbidding certain transformations).
I would say that if all your test cases go through with the simple definition of poison, stick with it. It's also consistent with the rest of LLVM semantics on poison. Select and phi are the only instructions at the moment that short-circuit poison.